### PR TITLE
Online payment feature - direct stripe integration

### PR DESCRIPTION
**Pull request by rishabhs95**
This PR is in succession to PR #1042.

This makes the feature SAQ-A PCI-compliant which is the easiest and cheapest amongst all available, thereby reducing the PCI burden for our clients.

screen shot 2016-07-27 at 8 36 32 pm

screen shot 2016-07-27 at 8 36 47 pm

screen shot 2016-07-27 at 8 37 47 pm

screen shot 2016-07-27 at 8 38 00 pm

### CODE REVIEW

File: app/controllers/payments_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    customer = Stripe::Customer.create(
+      :email => params[:stripeEmail],
+      :source  => params[:stripeToken]
+    )
+
+    gateway_response = Stripe::Charge.create(
+      :customer    => customer.id,
+      :amount      => @total_amount_to_pay.cents,
+      :description => 'Rails Stripe customer',
```
Comment (lagartoflojo) [inline app/controllers/payments_controller.rb]: We should probably improve this message :)

File: spec/factories/payments.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :payment do
+    user
+    conference
+    last4 4242
+    authorization_code 1234567890
```
Comment (lagartoflojo) [inline spec/factories/payments.rb]: since `authorization_code` is a string, better make this also a string.

File: spec/factories/payments.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :payment do
+    user
+    conference
+    last4 4242
```
Comment (lagartoflojo) [inline spec/factories/payments.rb]: since `last4` is a string, better make this also a string.

File: app/controllers/payments_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      :currency    => @conference.tickets.first.price_currency
+    )
+
+    payment = Payment.purchase(gateway_response, current_user, @conference)
```
Comment (lagartoflojo) [inline app/controllers/payments_controller.rb]: Why not handle this the way you did it for active merchant?
You can build a Payment with all the necessary data, and then call `payment.purchase` (no parameters) which will call Stripe and handle all that.

File: app/models/payment.rb [hunk 5]
```diff
@@ -6,8 +6,6 @@ class Payment < ActiveRecord::Base
   attr_accessor :stripeEmail
```
Comment (lagartoflojo) [inline app/models/payment.rb]: why is this camel-cased? looks veeery weird for ruby...

File: app/models/payment.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  def amount_to_pay
+    Ticket.total_price(conference, user, paid: false).cents
+  end
+
+  def purchase
+    customer = Stripe::Customer.create email: stripe_customer_email,
+                                       source: stripe_customer_token,
+                                       description: user.name
```
Comment (lagartoflojo) [inline app/models/payment.rb]: If the customer orders more than once, will this create the customer many times in Stripe? Would it be better to retrieve the customer if they already exist?

File: app/controllers/payments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    @total_amount_to_pay = Ticket.total_price(@conference, current_user, paid: false)
+    @unpaid_ticket_purchases = current_user.ticket_purchases.unpaid.by_conference(@conference)
+  end
+
+  def create
+    @total_amount_to_pay = Ticket.total_price(@conference, current_user, paid: false)
+
+    @payment = Payment.new payment_params.merge(stripe_customer_email: params[:stripeEmail],
+                                                stripe_customer_token: params[:stripeToken],
+                                                amount: @total_amount_to_pay.cents,
```
Comment (lagartoflojo) [inline app/controllers/payments_controller.rb]: Since you are calculating the amount to pay inside the Payment model, why do you need to pass it here?

File: app/models/payment.rb (hunk 6 above)
Comment (rishabhs95) [inline app/models/payment.rb]: Actually as we make the payment instantly, I think we can call Stripe's `Charge#create` directly instead of creating and saving customer details in our DB. 
I will remove new Customer creation part.

File: app/views/tickets/index.html.haml [hunk 8]
```diff
@@ -35,12 +35,11 @@
...
               %i.fa.fa-shopping-cart
-            = link_to 'Continue without a Ticket!', conference_conference_registration_path(@conference.short_title),
-            class: 'btn btn-danger btn-sm'
+            = link_to 'Cancel registration', conference_conference_registration_path(@conference.short_title), method: :delete, class: 'btn btn-danger btn-sm'
   .row
     .col-md-13
       %p.text-muted.text-center
         %small
-          * Buying a ticket is not mandatory. Checkout will be at the conference registration.
+          * Buying a ticket is mandatory. Your registration will not complete until you buy a ticket.
```
Comment (lagartoflojo) [inline app/views/tickets/index.html.haml]: In this screen, you can select no tickets and click "continue"...
Do you think we should validate that at least 1 ticket was chosen before continuing?

File: app/views/payments/_payment.html.haml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    %li
+      %a
+        = humanized_money_with_symbol @total_amount_to_pay
+  %script.stripe-button{ src: "https://checkout.stripe.com/checkout.js",
+        data: { amount: @total_amount_to_pay.cents,
+          currency: @total_amount_to_pay.currency,
+          image: image_url('OSEM_ICON.jpg'),
+          name: ENV['OSEM_NAME'] || 'OSEM',
+          description: "book your tickets",
+          key: Rails.application.secrets.stripe_publishable_key,
```
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: Here we're reading the stripe key from `secrets` instead of `configuration`, so configuring the keys with env variables actually doesn't work.

File: app/views/payments/_payment.html.haml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+          %td
+            = humanized_money_with_symbol ticket.quantity * ticket.price
+
+= form_tag conference_payments_path do
+  %ul.price-tags.pull-right
+    %li
+      %a
+        = humanized_money_with_symbol @total_amount_to_pay
+  %script.stripe-button{ src: "https://checkout.stripe.com/checkout.js",
+        data: { amount: @total_amount_to_pay.cents,
```
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: You can also specify the email here so that users don't have to type it in when the checkout modal opens.

File: app/views/payments/new.html.haml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+          %ul
+            - @payment.errors.full_messages.each do |msg|
+              %li= msg
+    .col-xs-8.col-xs-offset-2.well
+      = render partial: 'payment'
+  .row
+    .col-md-13
+      %p.text-muted.text-center
+        %small
+          The payment is totally secure. Your credit card details will be sent directly to our payment processor.
```
Comment (lagartoflojo) [inline app/views/payments/new.html.haml]: In American English, "totally" in this context sounds colloquial or even sarcastic.
I would say: 
> All payments are handled securely by our payment processor, [Stripe](https://stripe.com/about).

File: app/views/payments/_payment.html.haml (hunk 9 above)
Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: Yes, that's why I am telling users to put their key in `config/secrets.yml` in our documentation

File: app/views/payments/new.html.haml (hunk 11 above)
Comment (rishabhs95) [inline app/views/payments/new.html.haml]: Oh... okay :)

File: app/views/payments/_payment.html.haml (hunk 9 above)
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: But then why does the `stripe.rb` initializer try to read environment variables and set them in `Rails.configuration`?

Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: Using environment variables is, in general, a best practice anyway :)

http://www.clearlytech.com/2014/01/04/12-factor-apps-plain-english/#iclassicon-checkiiii.config---storeconfigintheenvironment

Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: Okay, I will check the API key like in initialiser.rb then :)

File: app/views/payments/_payment.html.haml (hunk 10 above)
Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: Okay, I've added that

File: app/views/payments/_payment.html.haml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+          currency: @total_amount_to_pay.currency,
+          image: image_url('OSEM_ICON.jpg'),
```
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: per the documentation PR https://github.com/openSUSE/osem/pull/1116, we should probably already be loading the image from the env variable, no?

Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: Also, the icon you uploaded seems unnecessarily big, considering the small size used in the checkout form.

File: config/initializers/stripe.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+Rails.configuration.stripe = {
+  :publishable_key => ENV['STRIPE_PUBLISHABLE_KEY'] || Rails.application.secrets.stripe_publishable_key,
+  :secret_key      => ENV['STRIPE_SECRET_KEY'] || Rails.application.secrets.stripe_secret_key
+}
```
Comment (lagartoflojo) [inline config/initializers/stripe.rb]: In `secrets.yml` you could just do 

```
  stripe_secret_key: <%= ENV["STRIPE_SECRET_KEY"] %>
```

and then you don't need this special code.
Then we can just use `Rails.secrets` directly, forget about `Rails.configuration`, and it will work with both `secrets.yml` and ENV vars.

File: app/controllers/payments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  def create
+    @payment = Payment.new payment_params.merge(stripe_customer_email: params[:stripeEmail],
+                                                stripe_customer_token: params[:stripeToken],
+                                                user: current_user,
+                                                conference: @conference)
+
+    if @payment.purchase && @payment.save
+      update_purchased_ticket_purchases
+      redirect_to conference_conference_registration_path(@conference.short_title), flash:
+        { success: 'Thanks! You have purchased your tickets successfully.' }
```
Comment (differentreality) [inline app/controllers/payments_controller.rb]: You should be able to use `success` instead of `flash` directly.

File: app/views/payments/_payment.html.haml (hunk 12 above)
Comment (differentreality) [inline app/views/payments/_payment.html.haml]: The icon is really cute, but please take into consideration that OSEM is not exclusively used by openSUSE. We should not force openSUSE logo here. The name of the file also seems to say OSEM_icon, but it is really openSUSE icon. We could use OSEM logo instead.

File: app/models/ability.rb [hunk 15]
```diff
@@ -81,6 +81,7 @@ def signed_in(user)
 
     can :index, Ticket
     can :manage, TicketPurchase, user_id: user.id
+    can [:new], Payment, user_id: user.id
```
Comment (differentreality) [inline app/models/ability.rb]: You should probably be authorizing `create` action too.

File: app/views/payments/_payment.html.haml (hunk 12 above)
Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: `OSEM_ICON` is the general name for icon used by our application for representing your organisation. As the organisation is openSUSE in the default case, so I've added openSUSE icon in the default case.

File: app/controllers/payments_controller.rb (hunk 14 above)
Comment (rishabhs95) [inline app/controllers/payments_controller.rb]: Okay, I've edited that

File: app/models/ability.rb (hunk 15 above)
Comment (rishabhs95) [inline app/models/ability.rb]: Ah, I will do that... thanks :)

File: app/views/payments/_payment.html.haml (hunk 12 above)
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: Maybe this line should say: `image: ENV['STRIPE_FORM_ICON']` (or something like that), so organizers can set it or not. If they don't, the checkout form won't show anything (and really, it doesn't HAVE to show anything).

File: app/views/tickets/index.html.haml (hunk 8 above)
Comment (rishabhs95) [inline app/views/tickets/index.html.haml]: Okay, I've added a check for that

File: app/views/payments/_payment.html.haml (hunk 12 above)
Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: I will remove the icon from it then

File: app/controllers/ticket_purchases_controller.rb [hunk 16]
```diff
@@ -8,9 +8,11 @@ def create
     message = TicketPurchase.purchase(@conference, current_user, params[:tickets][0])
     if message.blank?
       if current_user.ticket_purchases.by_conference(@conference).unpaid.any?
-        redirect_to new_conference_payment_path, notice: 'Please pay here to purchase tickets.'
+        redirect_to new_conference_payment_path,
+                    notice: 'Please pay here to purchase tickets.'
       else
-        redirect_to conference_conference_registration_path(@conference.short_title)
+        redirect_to conference_tickets_path(@conference.short_title),
+                    error: 'Please purchase atleast one ticket to continue.'
```
Comment (lagartoflojo) [inline app/controllers/ticket_purchases_controller.rb]: atleast => at least

File: config/initializers/stripe.rb [hunk 17]
```diff
@@ -1,6 +1 @@
-Rails.configuration.stripe = {
-  :publishable_key => ENV['STRIPE_PUBLISHABLE_KEY'] || Rails.application.secrets.stripe_publishable_key,
-  :secret_key      => ENV['STRIPE_SECRET_KEY'] || Rails.application.secrets.stripe_secret_key
-}
-
-Stripe.api_key = Rails.configuration.stripe[:secret_key]
+Stripe.api_key = Rails.application.secret.stripe_secret_key
```
Comment (lagartoflojo) [inline config/initializers/stripe.rb]: This works with env variables if you update secrets.yml as I mentioned in another comment.. will that change be in another commit?

File: config/secrets.yml.example [hunk 18]
```diff
@@ -64,5 +64,5 @@ production:
 
   # Register on stripe and add LIVE keys here
   # https://dashboard.stripe.com/account/apikeys
-  stripe_publishable_key: ''
-  stripe_secret_key: ''
```
Comment (lagartoflojo) [inline config/secrets.yml.example]: Please apply this change for the other environments as well :)

File: app/controllers/payments_controller.rb [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+  def new
+    @total_amount_to_pay = Ticket.total_price(@conference, current_user, paid: false)
+    @unpaid_ticket_purchases = current_user.ticket_purchases.unpaid.by_conference(@conference)
+  end
+
+  def create
+    @payment = Payment.new payment_params.merge(stripe_customer_email: params[:stripeEmail],
+                                                stripe_customer_token: params[:stripeToken],
+                                                user: current_user,
+                                                conference: @conference)
```
Comment (lagartoflojo) [inline app/controllers/payments_controller.rb]: I think if you move the `merge` call to the `payment_params` method, then this line will be cleaner :)
(It will be just `Payment.new payment_params`)

File: spec/models/payment_spec.rb [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+    let!(:conference) { create(:conference) }
+    let(:ticket_1) { create(:ticket, price: 10, price_currency: 'USD', conference: conference) }
+    let(:payment) { create(:payment, user: user, conference: conference) }
+
+    it ' returns correct unpaid amount' do
+      create(:ticket_purchase, ticket: ticket_1, user: user, quantity: 8)
+      expect(payment.amount_to_pay).to eq(8000)
+    end
+  end
+end
```
Comment (lagartoflojo) [inline spec/models/payment_spec.rb]: No tests for the `#purchase` method?

File: spec/models/payment_spec.rb [hunk 21]
```diff
@@ -32,4 +33,61 @@
...
+    end
+
+    context 'if the payment is not successful' do
+      before { StripeMock.prepare_card_error(:invalid_number) }
+
+      let(:payment) { create(:payment, user: user, conference: conference, stripe_customer_token: 'bogus_card_token', stripe_customer_email: user.email) }
+
+      before { payment.purchase }
+
+      context 'when the card is invalid' do
```
Comment (lagartoflojo) [inline spec/models/payment_spec.rb]: Can the gem also simulate a TimeOut? Are there any other errors we should consider?

File: app/models/payment.rb [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+                                             amount: amount_to_pay,
+                                             currency: conference.tickets.first.price_currency
+
+    self.amount = gateway_response[:amount]
+    self.last4 = gateway_response[:source][:last4]
+    self.authorization_code = gateway_response[:id]
+    self.status = 'success'
+    true
+
+  rescue => error
```
Comment (lagartoflojo) [inline app/models/payment.rb]: I recommend against rescuing from all types of errors.
For example, if we have some sort of exception from our own code (I dunno, division by zero or the Stripe gem changes its parameters, whatever), then users will see it as an error in their payment, and WE will never find out that something is wrong with our code.

It's better if we rescue from the exceptions that the Stripe gem can explicitly raise (invalid card? gateway error? timeout? ...?) and if there are any other errors, they will be sent to Errbit for analysis.

File: app/views/payments/_payment.html.haml [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+        = humanized_money_with_symbol @total_amount_to_pay
+  %script.stripe-button{ src: "https://checkout.stripe.com/checkout.js",
+        data: { amount: @total_amount_to_pay.cents,
+          email: current_user.email,
+          currency: @total_amount_to_pay.currency,
+          name: ENV['OSEM_NAME'] || 'OSEM',
+          description: "book your tickets",
+          key: Rails.application.secrets.stripe_publishable_key,
+          locale: "auto"}}
+  = link_to 'Edit Purchase', conference_tickets_path(@conference.short_title), class: 'btn btn-primary'
```
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: I'm not too excited about how this page looks. The "Pay with card" button looks nothing like the other buttons in the rest of OSEM, so it looks out of place. Plus we have two blue buttons. And there's the green ribbon which looks a bit out of place.. I know we are not designers 😂 but maybe we can improve it a bit. An idea:

**Before**
screen shot 2016-08-09 at 6 42 46 pm
 

**After**
screen shot 2016-08-09 at 6 42 29 pm

File: spec/features/ticket_purchases_spec.rb [hunk 24]
```diff
@@ -26,26 +26,12 @@
         fill_in "tickets__#{ticket.id}", with: '2'
         expect(current_path).to eq(conference_tickets_path(conference.short_title))
 
-        click_button 'Support'
+        click_button 'Continue'
 
+        expect(current_path).to eq(new_conference_payment_path(conference.short_title))
+        expect(flash).to eq('Please pay here to purchase tickets.')
         purchase = TicketPurchase.where(user_id: participant.id, ticket_id: ticket.id).first
         expect(purchase.quantity).to eq(2)
```
Comment (lagartoflojo) [inline spec/features/ticket_purchases_spec.rb]: Seems like this test is not finished.. the purchase process doesn't end here, does it?

File: spec/models/payment_spec.rb [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+        expect(payment.authorization_code).to eq('test_ch_3')
+      end
+    end
+
+    context 'if the payment is not successful' do
+      before { StripeMock.prepare_card_error(:invalid_number) }
+
+      let(:payment) { create(:payment, user: user, conference: conference, stripe_customer_token: 'bogus_card_token', stripe_customer_email: user.email) }
+
+      before { payment.purchase }
```
Comment (lagartoflojo) [inline spec/models/payment_spec.rb]: We could probably merge the two `before` blocks

File: spec/features/ticket_purchases_spec.rb (hunk 24 above)
Comment (rishabhs95) [inline spec/features/ticket_purchases_spec.rb]: Adding tests

File: spec/models/payment_spec.rb (hunk 25 above)
Comment (rishabhs95) [inline spec/models/payment_spec.rb]: Done.

File: app/views/payments/_payment.html.haml (hunk 23 above)
Comment (rishabhs95) [inline app/views/payments/_payment.html.haml]: Done.

File: spec/models/payment_spec.rb (hunk 21 above)
Comment (rishabhs95) [inline spec/models/payment_spec.rb]: Considering them

File: app/models/payment.rb (hunk 22 above)
Comment (rishabhs95) [inline app/models/payment.rb]: Okay

**Comment by rishabhs95** [comment]
We are testing our Payment class... I just added the line `expect{ Stripe::Charge.create }.to raise_error(...)` to show that an exception is raised but it is handled by `rescue`.

File: app/controllers/payments_controller.rb [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+  def create
+    @payment = Payment.new payment_params
+
+    if @payment.purchase && @payment.save
+      update_purchased_ticket_purchases
+      redirect_to conference_conference_registration_path(@conference.short_title),
+                  notice: 'Thanks! You have purchased your tickets successfully.'
+    else
+      @total_amount_to_pay = Ticket.total_price(@conference, current_user, paid: false)
```
Comment (ChrisBr) [inline app/controllers/payments_controller.rb]: Why don't we show the error in a flash message?
In Payment::purchase we add StripeError to ``errors.add(:base, error.message)`` so we should also propagate it to the users, shouldn't we?

File: app/controllers/ticket_purchases_controller.rb [hunk 27]
```diff
@@ -4,32 +4,22 @@ class TicketPurchasesController < ApplicationController
   authorize_resource :conference_registrations, class: Registration
 
   def create
+    TicketPurchase.by_conference(@conference).unpaid.by_user(current_user).destroy_all
```
Comment (ChrisBr) [inline app/controllers/ticket_purchases_controller.rb]: Why not 

``
current_user.ticket_purchases.by_conference(@conference).unpaid.destroy_all
``

File: app/views/conference_registrations/show.html.haml [hunk 28]
```diff
@@ -89,29 +89,27 @@
...
+                = tickets.first.price.symbol
+                = humanized_money tickets.first.price
+            %br
+          - if @tickets.any?
+            = link_to 'Buy more tickets', conference_tickets_path(@conference.short_title), class: "btn btn-default"
+          - else
+            You haven't bought any tickets.
+            = link_to 'Please buy some tickets to support us!', conference_tickets_path(@conference.short_title)
+            %p
+              (Your participation won't be valid without buying a ticket)
```
Comment (ChrisBr) [inline app/views/conference_registrations/show.html.haml]: Is this really true? I mean for instance we also have free conferences?

File: app/views/conference_registrations/show.html.haml [hunk 29]
```diff
@@ -89,29 +89,27 @@
           %span.fa-stack
             %i.fa.fa-square-o.fa-stack-2x
             %i.fa.fa-ticket.fa-stack-1x
-          Tickets
+          Ticket Purchases
```
Comment (ChrisBr) [inline app/views/conference_registrations/show.html.haml]: Shouldn't we move the heading also inside the ``tickets.any?``?
If there are no tickets, we don't need this section at all, do we?

Comment (lagartoflojo) [inline app/views/conference_registrations/show.html.haml]: Well we don't allow registrations with no tickets. We will allow free tickets to allow for this (https://github.com/openSUSE/osem/pull/1118).

Comment (ChrisBr) [inline app/views/conference_registrations/show.html.haml]: Ok

File: app/views/conference_registrations/show.html.haml (hunk 28 above)
Comment (ChrisBr) [inline app/views/conference_registrations/show.html.haml]: So this one is then also obsolete with https://github.com/openSUSE/osem/pull/1118, right?

Comment (lagartoflojo) [inline app/views/conference_registrations/show.html.haml]: yes :)

Comment (rishabhs95) [inline app/views/conference_registrations/show.html.haml]: I will just rebase #1118 over this branch, once it gets merged

File: app/controllers/ticket_purchases_controller.rb (hunk 27 above)
Comment (rishabhs95) [inline app/controllers/ticket_purchases_controller.rb]: okay, i have improved that

File: app/controllers/payments_controller.rb (hunk 26 above)
Comment (rishabhs95) [inline app/controllers/payments_controller.rb]: I do that in `app/views/payments/new.html.haml#7`

File: app/views/conference_registrations/show.html.haml (hunk 28 above)
Comment (differentreality) [inline app/views/conference_registrations/show.html.haml]: I had left a similar comment in the closed PR too.
I still don't like that phrase to exist here. It is not true, not even for this PR, because the registration to the conference is successful, and only after a successful registration we are redirected to the tickets page. I also find it very much against open source conferences, and this is the open source event manage!

Comment (lagartoflojo) [inline app/views/conference_registrations/show.html.haml]: The idea is that everyone has to get a ticket to come in the conference, even if the conference is free. In the PR #1118, we allow the creation of free tickets.
[Here](https://github.com/openSUSE/osem/pull/1108/files/0bc97925c6addad145afb8ddc830526b7b1e7a03#diff-3160d7d59726e8fa48c13ee0c12dcaf4R10) we check wether the user chose a ticket.

Considering the free-ticket scenario though, I think we should remove the mention of "purchase" or "buy" from the interface, and speak only about "getting" a ticket.

Comment (ChrisBr) [inline app/views/conference_registrations/show.html.haml]: +1 for @lagartoflojo suggestion!

Comment (rishabhs95) [inline app/views/conference_registrations/show.html.haml]: Then I think that a message like "Your participation won't be valid without getting a ticket" will be better here.

File: app/views/payments/_payment.html.haml [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+          %td
+            = ticket.quantity
+          %td
+            = humanized_money_with_symbol ticket.price
+          %td
+            = humanized_money_with_symbol ticket.quantity * ticket.price
+
+= form_tag conference_payments_path do
+  %script.stripe-button{ src: "https://checkout.stripe.com/checkout.js",
+        data: { amount: @total_amount_to_pay.cents,
```
Comment (lagartoflojo) [inline app/views/payments/_payment.html.haml]: These lines are too indented.

File: spec/support/external_request.rb [hunk 31]
```diff
@@ -1,6 +1,6 @@
 # Mock external requests to youtube
 require 'webmock/rspec'
-WebMock.disable_net_connect!(allow_localhost: true)
+WebMock.allow_net_connect!
```
Comment (lagartoflojo) [inline spec/support/external_request.rb]: We should not allow WebMock to connect to EVERYTHING, only to Stripe.

**Comment by lagartoflojo** [comment]
@differentreality @hennevogel @ChrisBr
Hi guys! We're ready to merge this.. do you have any additional comments before we do? Thanks!

**Comment by ChrisBr** [comment]
@lagartoflojo @rishabhs95 sorry, didn't have time to look at it!

However, congratulations and thanks for your good and continuous work this summer!
